### PR TITLE
Sasha/dynamo

### PR DESCRIPTION
**Pull request by klizhentas**
Update DynamoDB backend

* Add support for TTL
* Add support for Batch reads
* Update default values
* Use batch reads to retrieve nodes

### CODE REVIEW

File: lib/services/local/presence.go [hunk 1]
```diff
@@ -30,12 +31,20 @@ import (
 // PresenceService records and reports the presence of all components
 // of the cluster - Nodes, Proxies and SSH nodes
 type PresenceService struct {
+	*log.Entry
 	backend.Backend
+	// fast getter of the items extension
```
Comment (russjones) [inline lib/services/local/presence.go]: What about something like `getter is used to make batch requests to the backend.`

**Review by russjones — CHANGES_REQUESTED** [review]
Code looks good, documentation needs to be added for the Terraform modules.

File: lib/backend/dynamo/dynamodbbk.go [hunk 2]
```diff
@@ -628,17 +617,43 @@ func (b *DynamoDBBackend) GetVal(path []string, key string) ([]byte, error) {
...
-func suffix(key string) string {
-	vals := strings.Split(key, "/")
-	return vals[0]
+func convertError(err error) error {
+	if err == nil {
+		return nil
+	}
+	aerr, ok := err.(awserr.Error)
+	if !ok {
+		return err
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/backend/dynamo/dynamodbbk.go]: no wrap here intentionally to preserve original error

File: lib/backend/dynamo/dynamodbbk.go [hunk 3]
```diff
@@ -628,17 +617,43 @@ func (b *DynamoDBBackend) GetVal(path []string, key string) ([]byte, error) {
...
+	case dynamodb.ErrCodeProvisionedThroughputExceededException:
+		return trace.ConnectionProblem(aerr, aerr.Error())
+	case dynamodb.ErrCodeResourceNotFoundException:
+		return trace.NotFound(aerr.Error())
+	case dynamodb.ErrCodeItemCollectionSizeLimitExceededException:
+		return trace.BadParameter(aerr.Error())
+	case dynamodb.ErrCodeInternalServerError:
+		return trace.BadParameter(aerr.Error())
+	default:
+		return err
```
[the remark this answers is not in the document]
Comment (klizhentas) [reply lib/backend/dynamo/dynamodbbk.go]: same here

**Review by russjones — APPROVED** [review]

**Pull request merged**